client/homebrew/editor/stringArrayEditor/stringArrayEditor.jsx

```diff
@@ -109,35 +126,64 @@ const StringArrayEditor = createClass({
 		this.setState({ valueContext, updateValue: '' });
 	},
 
+	handleTagKeyDown : function(event, index) {
+		if(_.includes(['Enter', 'Space'], event.code)){
+			event.preventDefault();
+			this.editValue(index);
+			setTimeout(()=>{
+				event.target.querySelector('input').focus();
+			}, 0);
+		} else if(_.includes(['Delete'], event.key)){
+			this.removeValue(index);
+		} else if(_.includes(['ArrowLeft'], event.key)){
+			event.target.previousElementSibling?.focus();
+		} else if(_.includes(['ArrowRight'], event.key)){
+			event.target.nextElementSibling?.focus();
+		}
+
+
+	},
+
+	renderDatalist : function(){
+		if(this.props.options?.length > 0){
+			return <datalist id={`${this.props.id}__tags-precoordinated`}>
+				{this.props.options.map((option)=>{
+					return <option value={`${option}`}></option>;
+				})}
+			</datalist>;
+		}
+	},
+
 	render : function() {
-		const valueElements = Object.values(this.state.valueContext).map((context, i)=>context.editing
-			? <React.Fragment key={i}>
-				<div className='input-group'>
-					<input type='text' className={`value ${this.valueIsValid(this.state.updateValue, i) ? '' : 'invalid'}`} autoFocus placeholder={this.props.placeholder}
+		const valueElements = Object.values(this.state.valueContext).map((context, i)=>{
+			return <React.Fragment key={i}>
+				<div aria-label={`tag ${context.value}`} className={`tag ${context.editing ? 'editable' : ''}`} tabIndex={-1} onKeyDown={(e)=>this.handleTagKeyDown(e, i)}>
+					<span className={`tag-text ${context.editing ? 'hidden' : 'visible'}`} key={i} onClick={()=>this.editValue(i)}>{context.value}</span>
+					<input type='text' className={`value tag-input ${context.editing ? 'visible' : 'hidden'} ${this.valueIsValid(this.state.updateValue, i) || (this.state.updateValue == '') ? '' : 'invalid'}`} placeholder={this.props.placeholder}
 						value={this.state.updateValue}
 						onKeyDown={(e)=>this.handleValueInputKeyDown(e, i)}
-						onChange={(e)=>this.setState({ updateValue: e.target.value })}/>
-					{<div className='icon steel' onClick={(e)=>{ e.stopPropagation(); this.closeEditInput(i); }}><i className='fa fa-undo fa-fw'/></div>}
-					{this.valueIsValid(this.state.updateValue, i) ? <div className='icon steel'  onClick={(e)=>{ e.stopPropagation(); this.updateValue(this.state.updateValue, i); }}><i className='fa fa-check fa-fw'/></div> : null}
+						onChange={(e)=>this.setState({ updateValue: e.target.value })}
+						onBlur={()=>this.closeEditInput(i)}
+						list={this.props.options?.length > 0 ? `${this.props.id}__tags-precoordinated` : ''}/>
+					{this.renderDatalist()}
```

Comment:
The datalist is already rendered below under the `field` renderer, and this adds a separate copy for each tag. Does it need to be duplicated for each tag like this?

Reply:
No, we only need one `<datalist>` element. Elsewhere (another PR?) the matter of separating `<datalist>` as it's own component function has been discussed. (It could also be used for simple `<input>` elements too, not just StringArrayEditors. We don't have a specific need for that yet.)

Reply:
I'm pretty sure `client/components/combobox.jsx` that we use for the Languages selector is already this.